threads.go

```diff
@@ -187,6 +187,14 @@ func (t *threads) Close() (err error) {
 	if len(errs) > 0 {
 		return fmt.Errorf("failed while closing threads; err(s): %q", errs)
 	}
+
+	t.pullLock.Lock()
+	defer t.pullLock.Unlock()
+	// Wait for all thread pulls to finish
+	for _, semaph := range t.pullLocks {
+		semaph <- struct{}{}
+	}
```

Comment:
Quite important thing that was missing.
On `Close()`, we should wait for all thread pulls that might be running before saying we're done.

This basically tries to get all 1 buffered channels (semaphores) of threads and basically wait to _lock_ all of them. If that's the case, no thread pull is running.

Reply:
Clever!

Reply:
Yeah, basically using a 1 buffered channel is almost the same as a mutex. The thing is that this way you can test for lock and dismiss if locked. (with a default case).

I'm not sure I explained this decision originally. In `PullThread` we want to *try* to pull the thread, but if it's already being pulled we want to dismiss and go on. You can't "try" a Lock() with a mutex, but you can with a channel.

Anyway, this will make more sense in the next PR when these _semaphores_ are fully used (and delete the global lock). Did it now just not to forget and let the threadservice close when still are threads being pulled.